### DISCUSSION TITLE
UI Framework

### DISCUSSION THREAD

**Discussion by jondavid-black** [DevOps-MBSE/AaC#138] [Q&A]
We intentionally started simple with the built-in tkinter python UI framework...and we got a working MVP.  Although nobody really complained, I get the sense that the team really doesn't think tkinter is the right took for the job.  Should we do a trade study to evaluate options?

If you respond yet, please include candidates to consider and some ideas for evaluation criteria to evaluate within our use case.

**Comment by jondavid-black** [comment]
My initial thought (and why I posed the question) is...yes we should do a trade.

Potential Options:

1. Stick to a pure python solution:  tkinter, GTK, QT (PySide and PyQT), wxPython, Kivy
2. Stick with python but integrate a web-native front end:   
    - https://github.com/flexxui/flexx
    - https://www.fyears.org/2017/02/electron-as-gui-of-python-apps-updated.html
3. Abandon python and go all in on web-native:  https://github.com/electron/electron-quick-start

Potential Criteria:

1. License - Are the licenses of any integrated technologies consistent with AaC's open source license?
2. Testability - Do we know how we would test the solution in our CI/CD pipelines?
3. Complexity - Do we all need to get a PhD to comprehend the envisioned end state?
4. Maintainability - Would an AaC maintainer struggle to keep things working?  Is there any inherent fragility with the approach?
5. Rework - What (if any) rework would we have to plan for to make the UI usable?  Worst case may be a rewrite of AaC functionallity for the UI (which would be a hard no IMHO).
6. Rapid Design - Does the solution support rapid design iteration and evaluation?  Does it enable clean separation of concerns?
7. Capability - Does the solution enable graphical, tabluar, and visual AaC modeling?  Does it have out-of-the box features that ease the development of AaC model development and navigation tool?
8. Performance - When I click...do I wait?  How responsive is the solution assuming some reasonably large AaC model?
9. Extensibility- AaC uses a plugin approach for extensibility.  Is the solution compatible with the plug-in approach?  If AaC is extended in some way, does the UI solution extend with it somehow?
10. Curb Appeal - Is it (or can it be) pretty?  Poor visual appeal can put off a new user or make them feel AaC is low quality.
11. Cyber - Is the solution known to have security vulnerabilities?  Check for published vulnerabilities and see if/how they are addressed.  Ideally a maintained solution should be addressing any cyber security discoveries, but we need to check.

Trade Study Approach:
- We probably don't want to do an exhaustive trade.  If we do the trade, let's start by doing a little learning on various solutions and then voting for our top 2 to 4 to continue exploring.  The team has the flexibility to nominate candidate solutions not identified above.
- Of the candidate solutions we select through the voting process, let's replicate the MVP we did with tkinter in the last sprint.
- The developer of each MVP provides their assessment of criteria.  Others may provide an assessment as well but certain criteria should more heavily favor the feedback from the person who built the MVP than those who didn't (i.e. Maintainability).
- Capture our decision process within the AaC documentation.  We'll need to devise a good way to do this that we can reuse for future trade studies.  It's important to preserve why we make certain decisions.

**Reply by Coffee2Bits** [reply]
I like this approach for considering our available options.

I don't think that we need to pivot away from Python, but we could certainly consider using WebAssembly with popular languages like [C/C++, Go, Rust, Python, Java, etc.](https://stackoverflow.com/a/47483989) 

[An Introduction to Targeting Web Assembly (WASM) with Golang](https://www.youtube.com/watch?v=Jo7BbL7Xdms)
